src/biogeophys/BalanceCheckMod.F90

```diff
@@ -159,10 +204,98 @@ subroutine BeginWaterBalance(bounds, &
+    type(soilhydrology_type)   , intent(in)    :: soilhydrology_inst
+    type(lakestate_type)       , intent(in)    :: lakestate_inst
+    class(waterstate_type)     , intent(inout) :: waterstate_inst
+    class(waterflux_type)      , intent(inout) :: waterflux_inst
+    class(waterdiagnostic_type), intent(in)    :: waterdiagnostic_inst
+    class(waterbalance_type)   , intent(inout) :: waterbalance_inst
+    logical                    , intent(in)    :: use_aquifer_layer  ! whether an aquifer layer is used in this run
+    !
+    ! !LOCAL VARIABLES:
+    integer :: c, g, j, fc  ! indices
+    integer :: begc, endc, begg, endg  ! bounds
+    real(r8) :: qflx_liq_dynbal_left_to_dribble(bounds%begg:bounds%endg)  ! grc liq dynamic land cover change conversion runoff flux at beginning of time step
+    real(r8) :: qflx_ice_dynbal_left_to_dribble(bounds%begg:bounds%endg)  ! grc ice dynamic land cover change conversion runoff flux at beginning of time step
+    !-----------------------------------------------------------------------
+
+    associate(                                     &
+         zi        => col%zi                     , &  ! Input:  [real(r8) (:,:) ]  interface level below a "z" level (m)
+         zwt       => soilhydrology_inst%zwt_col , &  ! Input:  [real(r8) (:)   ]  water table depth (m)
+         aquifer_water_baseline => waterstate_inst%aquifer_water_baseline, &  ! Input: [real(r8)] baseline value for water in the unconfined aquifer (wa_col) for this bulk / tracer (mm)
+         wa        => waterstate_inst%wa_col     , &  ! Output: [real(r8) (:)   ]  water in the unconfined aquifer (mm)
+         begwb_col => waterbalance_inst%begwb_col, &  ! Output: [real(r8) (:)   ]  column-level water mass begining of the time step
+         begwb_grc => waterbalance_inst%begwb_grc  &  ! Output: [real(r8) (:)   ]  grid cell-level water mass begining of the time step
+         )
+
+    begc = bounds%begc
+    endc = bounds%endc
+    begg = bounds%begg
+    endg = bounds%endg
+
+    ! wa(c) gets added to liquid_mass in ComputeLiqIceMassNonLake
+    if(use_aquifer_layer) then
+       do fc = 1, num_nolakec
+          c = filter_nolakec(fc)
+          if (col%hydrologically_active(c)) then
+             if(zwt(c) <= zi(c,nlevsoi)) then
+                wa(c) = aquifer_water_baseline
+             end if
+          end if
+       end do
+    endif
+
+    ! NOTES subroutines Compute*Mass* are in TotalWaterAndHeatMod.F90
+    !       endwb is calculated in HydrologyDrainageMod & LakeHydrologyMod
+    call ComputeWaterMassNonLake(bounds, num_nolakec, filter_nolakec, &
+         waterstate_inst, waterdiagnostic_inst, &
+         subtract_dynbal_baselines = .false., &
+         water_mass = begwb_col(begc:endc))
+
+    call ComputeWaterMassLake(bounds, num_lakec, filter_lakec, &
+         waterstate_inst, lakestate_inst, &
+         add_lake_water_and_subtract_dynbal_baselines = .false., &
+         water_mass = begwb_col(begc:endc))
+
```

Comment:
Actually, as I think about this more: the main subtlety relates to the setting of wa(c) + aquifer_water_baseline. I forget exactly what this is doing – I remember this was a real subtlety that tripped me up a few years ago, and I know I have some notes about this that I could dig up if needed – but I'm thinking that this isn't directly related to balance checks, so should maybe be extracted to its own routine (like reset_wa) that is called early in the driver sequencing (exactly when may depend on when wa and zwt are set... my guess is that you'd want to do this before initializing the gridcell balance checks, but you may want to confirm that wa and zwt aren't changed between there and the initialization of the column balance checks). If you do that, then I think the remaining duplication (the two calls to ComputeWaterMass*) can stay as is - i.e., you don't need to do my above suggestion of introducing a shared routine.

Reply:
I commented out the `if(use_aquifer_layer)` code here (see https://github.com/ESCOMP/CTSM/pull/1228/commits/cb6a84042a47e785e365ac3629715bd22a5810d0) 
...and this test passed:
`ERP_D.f10_f10_musgs.IHistClm50Bgc.cheyenne_gnu.clm-decStart`
If the izumi test-suite also passes, I will move the `if(use_aquifer_layer)` code to its own routine as you suggested.

Reply:
Izumi test-suite OK (expected diffs from baseline, which for now is dev019 in my case).

Reply:
I just spent too much time trying to get my head back into this issue. I *think* this is okay the way you've done it now - just doing this reset in BeginWaterGridcellBalanceSingle. The one question I have is this: When we initialize a new column (due to dynamic landunits/columns), we set wa in the new column to wa in some template column on that grid cell. This initialization happens after the gridcell balance check initialization. With the old logic (where wa was reset to aquifer_water_baseline in the column-level balance check initialization), this resetting of wa happened after the dynamic landunit column initialization.

So my question is this... I think mainly for @swensosc -

- [ ] Is it a problem when use_aquifer_layer=true, if for one time step (the first timestep after initialization of a new column in a transient case) we have *not* reset wa(c) = aquifer_water_baseline? i.e., is it critical that wa(c) is reset to aquifer_water_baseline at the beginning of every time step when use_aquifer_layer = true, or is it okay for this not to be the case for one time step?

I *think* that, if the answer is that we need to do this resetting every time step, then the right place to do it is in dynInitColumnsMod:copy_state (I think that doing this resetting at the start of the column balance check, where it was done before, could potentially break the gridcell-level balance check).

Reply:
@billsacks I would benefit from a quick call to ensure that I'm on the same page as you about this. Prob. best to wait to hear from Sean first.

Reply:
@billsacks I am looking at subroutine copy_state. Let me know if my thoughts make sense:

I would reset wa with a subr. call just before the `end associate` statement. Subr. copy_state is already within a loop over `c`, so the new subroutine reset_water_in_unconf_aquifer would contain something like this (the do-loop removed):

```if(use_aquifer_layer) then
          if (col%hydrologically_active(c)) then
             if(zwt(c) <= zi(c,nlevsoi)) then
                wa(c) = aquifer_water_baseline
             end if
          end if
    endif```

The corresponding call from subr. BeginWaterGridcellBalanceSingle would look something like:

```do fc = 1, num_nolakec
          c = filter_nolakec(fc)
             call reset_water_in_unconf_aquifer(...)
   end do```

My first look suggests that I place subroutine reset_water_in_unconf_aquifer in WaterStateType.F90.

Anything else that I should be considering?

Reply:
I just got off a call with @swensosc . After going back and forth on this a bit, the end conclusion was that it's probably safe to skip doing this resetting in `copy_state`. The most likely place where this could cause any issues is here:

https://github.com/ESCOMP/CTSM/blob/0f6985da747a9e0538df63bf6e3f0ff6b4341229/src/biogeophys/SoilHydrologyMod.F90#L1261-L1266

but @swensosc feels that it's okay for things to be out of sync for one time step.

That said, I think your suggestion looks okay, though there are some performance-related issues: we generally try to avoid calling single-point routines in code that operates every time step (instead having the loop inside the subroutine)... there are ways we could do this, but it would be a little uglier... but we can talk about it if we want to go there.

**However, now that I'm writing this out, I think I see a solution I like better. This is inspired by a suggestion that @swensosc made that I didn't understand at the time, but now I see how this could work: Move the resetting of wa back into the start of the column-level balance check, and not the gridcell-level balance check, and track the loss of conservation due to this resetting.** Things I like about this are (1) it feels like the safest solution, because it is most consistent with how things currently work on master; (2) it explicitly tracks the non-conservation from this resetting, which someone might want to look at.

Specifically, I think this would involve:
- Removing the wa resetting code from the subroutine that initializes the gridcell-level balance check
- Reinstating this code in the subroutine that initializes the column-level balance check
- Creating a new column-level variable, maybe in waterbalance_type, like `wa_reset_nonconservation_gain_col`. This is the mass gained from this wa reset (negative indicates mass lost).
- Initialize that new variable to 0 for all columns in cold start initialization
- Calculate this variable as follows (this is the same as the existing wa resetting code, but with a few extra lines to calculate wa_reset_nonconservation_gain):
  ```Fortran
  if(use_aquifer_layer) then
     do fc = 1, num_nolakec
        c = filter_nolakec(fc)
        if (col%hydrologically_active(c)) then
           if(zwt(c) <= zi(c,nlevsoi)) then
              wa_reset_nonconservation_gain(c) = aquifer_water_baseline - wa(c)
              wa(c) = aquifer_water_baseline
           else
              wa_reset_nonconservation_gain(c) = 0._r8
           end if
        end if
     end do
  endif
  ```
- In the gridcell-level balance check, add this code:
  ```Fortran
  if (use_aquifer_layer) then
     ! wa_reset_nonconservation_gain_col should only be non-zero when use_aquifer_layer is
     ! true. We could do this c2g call always, but only do it when needed to avoid
     ! unnecessary calculations; also, by only adding this term when use_aquifer_layer is
     ! true, we effectively let the balance checks ensure that this term is zero when
     ! use_aquifer_layer is false, as it should be.
     call c2g( bounds,  &
          wa_reset_nonconservation_gain_col(bounds%begc:bounds%endc), &
          wa_reset_nonconservation_gain_grc(bounds%begg:bounds%endg), &
          c2l_scale_type= 'urbanf', l2g_scale_type='unity' )
  else
     wa_reset_nonconservation_gain_grc(bounds%begg:bounds%endg) = 0._r8
  end if
  ```
- Then subtract `wa_reset_nonconservation_gain_grc` from endwb_grc

@slevisconsulting does that seem like a reasonable solution to you?

Reply:
Yes. I will give it a try.

Reply:
@billsacks I included your recommended code modifications to the latest commit and confirmed that my usual test passes: `ERP_D.f10_f10_musgs.IHistClm50Bgc.cheyenne_gnu.clm-decStart`

I want to confirm that I grasp why this works. Pls check the following "hand-wavy" logic:
At the column level we haven't changed the balance check. We track the new non-conservation variable to account for it in the grid cell-level balance check. We do this in lieu of redetermining wa for begwb_grc and, therefore, must include the non-conservation term in the calculation of begwb_grc. All this pertains to the grid cell level due to changes at the column level caused by dynamic landuse. The non-conservation term (at the column and the grid cell level) accounts for column shifts from dynamic landuse because it is calculated in BeginWaterColumnBalanceSingle, which is after updating landuse.

@billsacks I would not mind including the above logic (with any corrections from you) as a comment in the code.

Reply:
@slevisconsulting I would give a different explanation than the one you gave. First off, while the dynamic landuse adjustment did play in to our discussion of this (and influenced our choice of how to implement this), the non-conservation term in the latest implementation really is unrelated to dynamic landuse, at least the way that I'm thinking about it: the adjustments in wa due to dynamic landuse are already accounted for by the dynbal dribblers. Instead, I would explain this as follows:

We want to adjust wa back to the baseline under certain conditions. The right way to do this might be to use explicit fluxes from some other state, but in this case we don't have a source to pull from, so we just adjust wa without any explicit fluxes. Because we do this adjustment before initializing the column-level balance check, the column-level check is completely unaware of the adjustment. However, since this adjustment happens after initializing the gridcell-level balance check, we *do* have to account for this adjustment in the gridcell-level balance check. The normal way to account for an adjustment like this would be to include the flux in the balance check. But here we don't have an explicit flux, so instead track the non-conservation *state*. (In principle, we could calculate an explicit flux and use that, but we don't gain anything from using an explicit flux in this case.)

(This doesn't explain *why* we chose to use this implementation. That explanation *would* tie-in with subtleties of dynamic land use. I can take a stab at explaining that, too, if you'd like.)

Reply:
I included your explanation in the latest commit. Thanks, @billsacks 

Explaining the subtleties of dynamic landuse that come into play seems like a good idea for when we look at this again later. I for one will not remember these subtleties even a couple of months from now :-)

In the explanation, I am interested in understanding why it makes sense to place the code below at the initialization rather than the finalizing of the grc-level balance check:

```
if (use_aquifer_layer) then
       call c2g( bounds, &
            wa_reset_nonconservation_gain_col(begc:endc), &
            wa_reset_nonconservation_gain_grc(begg:endg), &
            c2l_scale_type='urbanf', l2g_scale_type='unity' )
else
       wa_reset_nonconservation_gain_grc(begg:endg) = 0._r8
end if

[...]

    do g = begg, endg
       begwb_grc(g) = begwb_grc(g) - qflx_liq_dynbal_left_to_dribble(g)  &
                                   - qflx_ice_dynbal_left_to_dribble(g)  &
                                   - wa_reset_nonconservation_gain_grc(g)
    end do
```

Reply:
To be abs. clear, if this were moved to the finalizing of the grc-level bal. check, the sign in front of wa_reset_... would change to `+`. 

So again, does this code need to be run at initialization instead of when finalizing? If either is ok, I'm wondering whether it is more intuitive at the finalizing because the wa_reset_... term is from the same time step.

Reply:
- [ ] Ah, yes, sorry: I missed that when reviewing your latest changes. Indeed, I was imagining that wa_reset_nonconservation_gain_grc would adjust endwb_grc, not begwb_grc. I think you want to subtract it from endwb_grc. (Imagine that wa was 4000 at the start of the time step, and the baseline is 5000. Then wa would have been increased by 1000, and wa_reset_nonconservation_gain will also be 1000. To keep the balance check working, I think we need to subtract this 1000 from endwb... if you feel otherwise, please say so, because I'm not 100% sure about this.)

If you want to say something about the relationship to dynamic landuse, I would probably put a comment like this in BeginWaterColumnBalanceSingle: This resetting of wa needs to be done after dynamic landuse so that it is done after initializing any new columns; therefore, we cannot do this resetting before initializing the gridcell-level balance check (since that is done before dynamic landuse). We do it here before initializing the column-level balance check so that we don't need to account for this non-conservation in the column-level balance check, although we still need to account for this non-conservation in the gridcell-level balance check.

Reply:
This makes sense, and I apologize for misunderstanding the first time around. Now I am less inclined to include this most recent explanation, so I will leave it out.